### PR TITLE
Authenticate via google

### PR DESCRIPTION
**Pull request by daothanhcam**

### CODE REVIEW

File: app/controllers/omniauth_callbacks_controller.rb [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    else
+      session["devise.#{provider}_data"] = env["omniauth.auth"]
+      redirect_to new_user_registration_url
+    end
+  end
+
+  def upgrade
+    scope = nil
+    if params[:provider] == "google_oauth2"
+      scope = "email,profile,offline,https://www.googleapis.com/auth/admin.directory.user"
```
Comment (phuong1492) [inline app/controllers/omniauth_callbacks_controller.rb]: 1 space after ","

File: app/controllers/omniauth_callbacks_controller.rb [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    end
+    redirect_to user_omniauth_authorize_path(params[:provider]),
+      flash: {scope: scope}
+  end
+
+  def setup
+    request.env["omniauth.strategy"].options["scope"] = flash[:scope] ||
+      request.env["omniauth.strategy"].options["scope"]
+    render :text => "Setup complete.", :status => 404
```
Comment (phuong1492) [inline app/controllers/omniauth_callbacks_controller.rb]: status: 404?

File: app/models/form_user.rb [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+                          allow_blank: true, if: :email_changed?
+
+  validates_presence_of     :password, if: :password_required?
+  validates_confirmation_of :password, if: :password_required?
+  validates_length_of       :password, within: Devise.password_length,
+                            allow_blank: true
+
+  def password_required?
+    return false if email.blank?
+    !persisted? || !password.nil? || !password_confirmation.nil?
```
Comment (phuong1492) [inline app/models/form_user.rb]: ident

File: app/models/user.rb [hunk 4]
```diff
@@ -1,9 +1,26 @@
...
+
   has_many :addresses, dependent: :destroy
   has_many :reviews, dependent: :destroy
   has_many :comments, dependent: :destroy
+  has_many :identities
 
   enum role: [:normal, :admin]
+
+  def google_oauth2
+    identities.where(:provider => "google_oauth2").first
```
Comment (phuong1492) [inline app/models/user.rb]: provider: "google_oauth2"

File: app/controllers/omniauth_callbacks_controller.rb (hunk 2 above)
Comment (ThuBM) [inline app/controllers/omniauth_callbacks_controller.rb]: I18n

File: Gemfile [hunk 5]
```diff
@@ -1,5 +1,6 @@
 source "https://rubygems.org"
 
+gem "dotenv-rails", :groups => [:development, :test]
```
Comment (NguyenHai2310) [inline Gemfile]: use groups: []

File: app/controllers/omniauth_callbacks_controller.rb [hunk 6]
```diff
@@ -0,0 +1,46 @@
+class OmniauthCallbacksController < Devise::OmniauthCallbacksController
+  def google_oauth2
+    generic_callback("google_oauth2")
+  end
+
+  def generic_callback(provider)
```
Comment (NguyenHai2310) [inline app/controllers/omniauth_callbacks_controller.rb]: try remove ()

File: app/models/identity.rb [hunk 7]
```diff
@@ -0,0 +1,18 @@
+class Identity < ActiveRecord::Base
+  belongs_to :user
+  validates_presence_of :uid, :provider
+  validates_uniqueness_of :uid, :scope => :provider
```
Comment (NguyenHai2310) [inline app/models/identity.rb]: use scope: :provide

File: app/models/identity.rb [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+
+  def self.find_for_oauth(auth)
```
Comment (NguyenHai2310) [inline app/models/identity.rb]: try remove ()

File: app/controllers/omniauth_callbacks_controller.rb [hunk 9]
```diff
@@ -0,0 +1,46 @@
+class OmniauthCallbacksController < Devise::OmniauthCallbacksController
+  def google_oauth2
+    generic_callback("google_oauth2")
```
Comment (phuong1492) [inline app/controllers/omniauth_callbacks_controller.rb]: can remove () here?

File: app/controllers/omniauth_callbacks_controller.rb [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+      end
+    else
+      session["devise.#{provider}_data"] = env["omniauth.auth"]
+      redirect_to new_user_registration_url
+    end
+  end
+
+  def upgrade
+    scope = nil
+    if params[:provider] == "google_oauth2"
```
Comment (phuong1492) [inline app/controllers/omniauth_callbacks_controller.rb]: I18n?

File: app/controllers/registration_controller.rb [hunk 11]
```diff
@@ -0,0 +1,17 @@
+class RegistrationsController < Devise::RegistrationsController
+  def update_resource(resource, params)
+    if resource.encrypted_password.blank?
+      resource.email = params[:email] if params[:email]
+      unless params[:password].blank? && params[:password] == params[:password_confirmation]
+        logger.info "Updating password"
```
Comment (phuong1492) [inline app/controllers/registration_controller.rb]: I18n?

File: app/controllers/omniauth_callbacks_controller.rb [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+      scope = "email, profile, offline, https://www.googleapis.com/auth/admin.directory.user"
+    end
+    redirect_to user_omniauth_authorize_path(params[:provider]),
+      flash: {scope: scope}
+  end
+
+  def setup
+    request.env["omniauth.strategy"].options["scope"] = flash[:scope] ||
+      request.env["omniauth.strategy"].options["scope"]
+    render text: t("Setup complete."), status: t "404"
```
Comment (phuong1492) [inline app/controllers/omniauth_callbacks_controller.rb]: Can remove () ?

File: app/controllers/registration_controller.rb [hunk 13]
```diff
@@ -0,0 +1,17 @@
+class RegistrationsController < Devise::RegistrationsController
+  def update_resource(resource, params)
```
Comment (phuong1492) [inline app/controllers/registration_controller.rb]: remove ()?

File: app/controllers/registration_controller.rb [hunk 14]
```diff
@@ -0,0 +1,17 @@
...
+    if resource.encrypted_password.blank?
+      resource.email = params[:email] if params[:email]
+      unless params[:password].blank? && params[:password] == params[:password_confirmation]
+        logger.info "Updating password"
+        resource.password = params[:password]
+        resource.save
+      end
+      if resource.valid?
+        resource.update_without_password(params)
```
Comment (phuong1492) [inline app/controllers/registration_controller.rb]: remove()

File: app/controllers/registration_controller.rb [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+      end
+    else
+      resource.update_with_password(params)
```
Comment (phuong1492) [inline app/controllers/registration_controller.rb]: remove()

File: app/models/user.rb [hunk 16]
```diff
@@ -1,9 +1,26 @@
...
   enum role: [:normal, :admin]
+
+  def google_oauth2
+    identities.where(provider: "google_oauth2").first
+  end
+
+  def google_oauth2_client
+    unless @google_oauth2_client
+      @google_oauth2_client = Google::APIClient
+        .new(application_name: "omniauth-google-oauth2")
```
Comment (phuong1492) [inline app/models/user.rb]: can remove()?

File: app/controllers/omniauth_callbacks_controller.rb [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+  def upgrade
+    if params[:provider] == Settings.provider.google
+      scope = "email, profile, offline, https://www.googleapis.com/auth/admin.directory.user"
+    end
+    redirect_to user_omniauth_authorize_path params[:provider],
+      flash: {scope: scope}
+  end
+
+  def setup
+      request.env["omniauth.strategy"].options["scope"] ||= flash[:scope]
```
Comment (ThuBM) [inline app/controllers/omniauth_callbacks_controller.rb]: indent

File: app/views/users/shared/_links.html.erb [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+	<br />
+	<%- if controller_name != "sessions" %>
+	  <%= link_to t("login"), new_session_path(resource_name) %><br />
+	<% end -%>
+
+	<%- if devise_mapping.registerable? && controller_name != "registrations" %>
+	  <%= link_to t("sign_up"), new_registration_path(resource_name) %><br />
+	<% end -%>
+
+	<%- if devise_mapping.recoverable? && controller_name != "passwords" && controller_name != "registrations" %>
```
Comment (ThuBM) [inline app/views/users/shared/_links.html.erb]: Line too long

File: app/views/users/shared/_links.html.erb [hunk 19]
```diff
@@ -0,0 +1,30 @@
...
+	  	<%= link_to t("forgot_password"), new_password_path(resource_name) %><br />
+	<% end -%>
+
+	<%- if devise_mapping.confirmable? && controller_name != "confirmations" %>
+	  <%= link_to t("confirm_instructions"),
+	    new_confirmation_path(resource_name) %><br />
+	<% end -%>
+
+	<%- if devise_mapping.lockable? && resource_class.unlock_strategy_enabled?(:email) && controller_name != "unlocks" %>
```
Comment (ThuBM) [inline app/views/users/shared/_links.html.erb]: line too long

File: app/models/form_user.rb [hunk 20]
```diff
@@ -0,0 +1,22 @@
...
+                          allow_blank: true, if: :email_changed?
+
+  validates_presence_of     :password, if: :password_required?
+  validates_confirmation_of :password, if: :password_required?
+  validates_length_of       :password, within: Devise.password_length,
+                            allow_blank: true
+
+  def password_required?
+    return false if email.blank?
+      !persisted? || !password.nil? || !password_confirmation.nil?
```
Comment (ThuBM) [inline app/models/form_user.rb]: indent

File: app/controllers/registration_controller.rb [hunk 21]
```diff
@@ -0,0 +1,18 @@
+class RegistrationsController < Devise::RegistrationsController
+  def update_resource resource, params
+    if resource.encrypted_password.blank?
+      if params[:email]
+        resource.email = params[:email]
+      unless params[:password].blank? || params[:password] !=
```
Comment (DucNguyenVan) [inline app/controllers/registration_controller.rb]: i think i should refactor by De Morgan Law

File: app/models/form_user.rb (hunk 3 above)
Comment (DucNguyenVan) [inline app/models/form_user.rb]: use De Morgan Law too

File: app/controllers/registration_controller.rb (hunk 21 above)
Comment (daothanhcam) [inline app/controllers/registration_controller.rb]: I'm trying do not use !

Comment (DucNguyenVan) [inline app/controllers/registration_controller.rb]: yes, i can code like:
if params[:password].present? && params[:password] == params[:password_confirmation]

File: app/models/form_user.rb [hunk 22]
```diff
@@ -0,0 +1,22 @@
...
+                          allow_blank: true, if: :email_changed?
+
+  validates_presence_of     :password, if: :password_required?
+  validates_confirmation_of :password, if: :password_required?
+  validates_length_of       :password, within: Devise.password_length,
+                            allow_blank: true
+
+  def password_required?
+    return false if email.blank?
+    persisted? && password.nil? && password_confirmation.nil?
```
Comment (DucNguyenVan) [inline app/models/form_user.rb]: !(persisted? && password.nil? && password_confirmation.nil?)

File: app/controllers/registration_controller.rb [hunk 23]
```diff
@@ -0,0 +1,18 @@
+class RegistrationsController < Devise::RegistrationsController
+  def update_resource resource, params
+    if resource.encrypted_password.blank?
+      if params[:email]
```
Comment (hieuns) [inline app/controllers/registration_controller.rb]: Chỗ này cậu nên chuyển thành resource.email = params[:emal] if params[:email] cho gọn.

File: app/controllers/registration_controller.rb (hunk 21 above)
Comment (hieuns) [inline app/controllers/registration_controller.rb]: Cậu có thể chuyển thành if params[:password].present? && params[:password] == params[:password_confirmation]

File: app/views/users/shared/_links.html.erb [hunk 24]
```diff
@@ -0,0 +1,35 @@
+<div class="col-md-offset-3">
+  <br />
+  <% if controller_name != "sessions" %>
+    <%= link_to t("login"), new_session_path(resource_name) %><br />
```
Comment (DucNguyenVan) [inline app/views/users/shared/_links.html.erb]: you can change to be one-line with "if' condition operator

File: app/views/users/shared/_links.html.erb [hunk 25]
```diff
@@ -0,0 +1,35 @@
...
+  <% end %>
+
+  <% if devise_mapping.registerable? && controller_name != "registrations" %>
```
Comment (DucNguyenVan) [inline app/views/users/shared/_links.html.erb]: refactor code from line 7 to line 15
you can combine two "registration" condition

**Pull request merged**